controllers/managedfusiondeployment_controller.go

```diff
@@ -0,0 +1,795 @@
+		r.k8sMetricsServiceMonitor.Spec = desired.Spec
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("Failed to update k8sMetricsServiceMonitor: %v", err)
+	}
+	return nil
+}
+
+// reconcileMonitoringResources labels all monitoring resources (ServiceMonitors, PodMonitors, and PrometheusRules)
+// found in the target namespace with a label that matches the label selector the defined on the Prometheus resource
+// we are reconciling in reconcilePrometheus. Doing so instructs the Prometheus instance to notice and react to these labeled
+// monitoring resources
+func (r *ManagedFusionDeployment) reconcileMonitoringResources() error {
+	r.Log.Info("reconciling monitoring resources")
+
+	podMonitorList := promv1.PodMonitorList{}
+	if err := r.list(&podMonitorList); err != nil {
+		return fmt.Errorf("Could not list pod monitors: %v", err)
+	}
+	for i := range podMonitorList.Items {
+		obj := podMonitorList.Items[i]
+		utils.AddLabel(obj, monLabelKey, monLabelValue)
+		if err := r.update(obj); err != nil {
+			return err
+		}
+	}
+
+	serviceMonitorList := promv1.ServiceMonitorList{}
+	if err := r.list(&serviceMonitorList); err != nil {
+		return fmt.Errorf("Could not list service monitors: %v", err)
+	}
+	for i := range serviceMonitorList.Items {
+		obj := serviceMonitorList.Items[i]
+		utils.AddLabel(obj, monLabelKey, monLabelValue)
+		if err := r.update(obj); err != nil {
+			return err
+		}
+	}
+
+	promRuleList := promv1.PrometheusRuleList{}
+	if err := r.list(&promRuleList); err != nil {
+		return fmt.Errorf("Could not list prometheus rules: %v", err)
+	}
+	for i := range promRuleList.Items {
+		obj := promRuleList.Items[i]
+		utils.AddLabel(obj, monLabelKey, monLabelValue)
+		if err := r.update(obj); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+func (r *ManagedFusionDeployment) reconcilePrometheusProxyNetworkPolicy() error {
+	r.Log.Info("reconciling PrometheusProxyNetworkPolicy resources")
+
+	_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, r.prometheusProxyNetworkPolicy, func() error {
+		if err := r.own(r.prometheusProxyNetworkPolicy); err != nil {
+			return err
+		}
+		desired := templates.PrometheusProxyNetworkPolicyTemplate.DeepCopy()
+		r.prometheusProxyNetworkPolicy.Spec = desired.Spec
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("Failed to update prometheus proxy NetworkPolicy: %v", err)
+	}
+	return nil
+}
+
+func (r *ManagedFusionDeployment) initiateAgentUninstallation() error {
+	r.Log.Info("deleting agent namespace")
+	if err := r.delete(&corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: r.Namespace,
+		},
+	}); err != nil {
+		return fmt.Errorf("failed to delete namespace: %v", err)
+	}
+	r.Log.Info("Agent namespace deleted successfully")
+
+	r.Log.Info("deleting agent csv")
+	if err := r.deleteCSVByPrefix(agentCSVPrefix); err != nil {
+		return fmt.Errorf("Unable to delete csv: %v", err)
+	}
+	r.Log.Info("Agent csv removed successfully")
+	return nil
```

Comment:
Can we even do that in this order? will we be able to remove resources in the namespace after namespace deletion command? 

Also, what about the subscription and the operator group? Do we need to even delete any of these if we are just deleting the namespace? And happens to the operator pods?

Reply:
Removing the namespace removes everything(includes subscription and operator group) so we are just initiating namespace deletion.

Reply:
Updated in latest commit.